#638, Settings/pins/hidden posts/filters etc. import/export

K1rakishou opened the pull request:
Closes #638, #289 and couple of bugs #674 and #677.
This is a big boy and needs good testing.

Adamantcheese commented:
Two things:
1) [It appears there's already a restart function call in StartActivity, is there any particular reason you need to have a new one that looks very similar to this one? It's used for when the Theme changes.](https://github.com/Floens/Clover/blob/c8cdf29d030bb37922a9d8492c667e60c6ac08de/Clover/app/src/main/java/org/floens/chan/ui/activity/StartActivity.java#L306)
2) When opening up the import/export settings dialog for the first time, it doesn't prompt for getting the permission. This causes it to fail to make the proper folder and even if you grant access in the next step to export, it'll fail to export. If you exit out of the menu and go back into import/export again, it'll work correctly.

K1rakishou commented:
@Adamantcheese 
1. IIRC it didn't work for me when I tried it but maybe it was because I did something wrong. Alright, I will try that method again an if it works I will remove the one I've added.
2. Yes, this is a problem and I'll fix it.
3. I didn't just remove them, actually I moved the dependency initialization here `Clover/app/src/main/java/org/floens/chan/core/di/`
I was getting some strange exception (Feather cannot inject databaseManager to the Chan class) so I moved dependencies around until they all ended up in one place (The exception was actually because I had removed the default constructor from the SiteModel but Feather absorbs the stacktrace so it wasn't obvious at all). I can revert that change, but I personally think it is better when all dependencies are being initialized in Modules.
4. Yes, I did forget that you can just read them all with getAll(). I'll change it.

K1rakishou commented:
@Adamantcheese 
1. So the other restart method didn't work for me.
2. Fixed.
3. Didn't change anything.
4. Looks like it is impossible to deserialize shared prefs from a string without enumerating the whole collection of settings because you have to know at least what type the value is. I mean it's pretty easy to serialize them to a string (just check what instance is the value) but kinda hard to deserialize. Primarily because some settings contain numbers but their type is StringSetting (like this one: `public static final StringSetting fontSize`). So I have to leave it like it is now (with huge  list of all possible settings).

K1rakishou commented:
Oh, wait a  minute. It looks like I can just read the whole file directly without using the SharedPrefs class and the just write it back the same way. I'm gonna experiment with this tomorrow.

Adamantcheese commented:
Cool, thanks for all the updates. I'll wait another little bit to merge it in, but everything looks fine. I'll test it properly when you do that last change for 4.

K1rakishou commented:
@Adamantcheese 
Alright, done. Now there is only one problem left when adding/deleting columns in the database do not forget to handle that in ImportExportRepository onUpgrade/onDowngrade methods. Otherwise it may broke everything for the user. The settings file also has it's own version (`private int version = CURRENT_EXPORT_SETTINGS_VERSION`;). So don't forget to change it as well when changing something in the export schema. Sometimes it will not be possible to restore some settings from the file when upgrading/downgrading. Like when I added "whole_thread" column to the ThreadHide table. It is impossible to figure out this parameter on it's own so the only thing that can be done here is to completely clear all the ThreadHides. I have no idea how to mitigate this issue. The only thing is probably not change the database columns.

Adamantcheese commented:
Adding a comment at the end of DatabaseHelper's onUpgrade method to say that ImportExportRepository's onUpgrade method should be modified if needed if onUpgrade changes should be enough warning for other devs.
I don't believe the onDowngrade method will ever be used though. The database only has an onUpgrade method because the database version should only ever go up. The same should apply to the export settings version. It should probably be removed to avoid confusion.
And now that I'm looking at it, you parse the file with the fromJson(String, Class) method. If the exported class changes (or any nested classes) and you try to use that, I think it'll throw an exception and not allow an import? It should probably log "Error while trying to import pins", which isn't even a correct error message for what either the import or export functions do anymore.

K1rakishou commented:
@Adamantcheese 
>Adding a comment at the end of DatabaseHelper's onUpgrade method to say that ImportExportRepository's onUpgrade method should be modified if needed if onUpgrade changes should be enough warning for other devs.

I agree. Will add it.

>I don't believe the onDowngrade method will ever be used though. The database only has an onUpgrade method because the database version should only ever go up. The same should apply to the export settings version. It should probably be removed to avoid confusion.

Yeah, probably it won't be used. So in case of downgrade we can just return empty settings (`ExportedAppSettings.empty()`) which will just do nothing. But I don't think it is a good idea to remove the version. This way users won't be able to migrate their settings from one version to another when something in the database changes (or it may be even worse, like, the settings will apply without exception but it will fuck up something else and user won't be able to use the app). 

>And now that I'm looking at it, you parse the file with the fromJson(String, Class) method. If the exported class changes (or any nested classes) and you try to use that, I think it'll throw an exception and not allow an import? 

No it will just use a default value (like 0 for ints or null for strings or any other objects).
![345345](https://user-images.githubusercontent.com/10332458/56457253-787e0800-6380-11e9-9fae-bd837b0c409f.png)
And that's why we need the version and onUpgrade method where you can do something with data that couldn't be imported due to changes.

Adamantcheese commented:
I've taken a good amount of time to look at it, and I can't see any issues anymore. I've merged it in with my fork.

K1rakishou commented:
@Adamantcheese Thanks.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/controller/export/ImportExportSettingsController.java:
```diff
@@ -0,0 +1,306 @@
+
+        getPermissionHelper().requestPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, granted -> {
+            if (granted) {
+                navigationController.presentController(loadingViewController);
```
Probably should move this inside the check below because if the presenter is not set then we will show the loading view but won't start the exporing so it will just be shown forever.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/controller/export/ImportExportSettingsController.java:
```diff
@@ -0,0 +1,306 @@
+    }
+
+    private void onStartImportButtonClicked() {
+        navigationController.presentController(loadingViewController);
```
Same as above. This needs to be moved inside the check below.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/ui/controller/export/LoadingViewController.java:
```diff
@@ -0,0 +1,50 @@
+import org.floens.chan.controller.Controller;
+import org.floens.chan.utils.AndroidUtils;
+
+public class LoadingViewController extends Controller {
```
Like 99% of this class code can be extracted to an abstract class and then other controllers that use dark transparent background can inherit from it.

K1rakishou commented on Clover/app/src/main/res/values/strings.xml:
```diff
@@ -577,4 +589,20 @@ Don't have a 4chan Pass?<br>
+    <string name="import_settings" tools:ignore="MissingTranslation">Import settings</string>
+    <string name="import_settings_from_a_file" tools:ignore="MissingTranslation">Import settings from a file</string>
+    <string name="import_warning_title" tools:ignore="MissingTranslation">Warning</string>
+    <string name="import_warning_text" tools:ignore="MissingTranslation">You are about to import settings/saved pins/filters/hidden posts from a file. The file must be located at \"%1$s\" and have a name \"%2$s\". This will clear all your existing settings/pins/threads and replace them with the ones from the file. You MAY LOOSE some of your settings/pins/hidden threads if your are trying to impor ...
```
Correct this message.

Adamantcheese commented on Clover/app/src/main/java/org/floens/chan/core/repository/ImportExportRepository.java:
```diff
@@ -0,0 +1,476 @@
+                    );
+                }
+
+                try (RandomAccessFile raf = new RandomAccessFile(settingsFile, "rw")) {
```
This doesn't work correctly with UTF-8 encoded strings. Some symbols tend to get fucked up and then you won't be able to import afterwards. Use a FileWriter instead, which writes out UTF-8 by default and doesn't cause issues later on. Noticeably, ™ seems to turn into " for some reason?

K1rakishou replied:
Yeah, apparently for UTF to work it should be writeUtf() not writeBytes().

K1rakishou commented:
Should be fixed now. Tested on bookmarked threads with japanese characters in the title.
Btw, I have changed the CURRENT_EXPORT_SETTINGS_VERSION to version 2 and disabled upgrading from v1 to v2 because it will be impossible to deserialize v1 settings on v2. So when releasing a new version of the app we probably should notify the users (in the changelist) that they should re-export their settings again.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/presenter/ImportExportSettingsPresenter.java:
```diff
@@ -40,80 +42,89 @@
-                if (callbacks != null) {
-                    callbacks.onError("Error while trying to export settings = " + error.getMessage());
+                if (exportCallbacks != null) {
+                    exportCallbacks.onError("Error while try to export settings: " + error.getMessage());
```
"trying".

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/repository/ImportExportRepository.java:
```diff
@@ -60,7 +59,7 @@
 
     // Don't forget to change this when changing any of the Export models.
     // Also, don't forget to handle the change in the onUpgrade or onDowngrade methods
```
Remove "or onDowngrade" since it is not supported.

K1rakishou commented on Clover/app/src/main/java/org/floens/chan/core/repository/ImportExportRepository.java:
```diff
@@ -295,7 +294,20 @@ private void writeSettingsToDatabase(@NonNull ExportedAppSettings appSettingsPar
+            // Upgrade from version 1 to version 2.
+            // In version 2 we had to change the way the settings are serialized to a file because
+            // it was not considering the UTF-8 characters. This breaks everything. So we have no other
+            // way other than clear everything.
```
"other than not import anything"

Adamantcheese replied:
This is entirely unnecessary because the exception being thrown will be in line 151. The actual JSON fails to parse because of mismatched quotes, so you'd never even get to this point. There's no need to increment the version number.

K1rakishou commented:
Yeah, that's a problem. I think I will have to rewrite the structure of the ExportedSettings so that the version field is not a part of the json. Because right now we need to read the json just to get a version. And because of that we can't check the file version before we start parsing. I don't want to leave it as it is because the error message in this case is not really descriptive to the user.

K1rakishou commented:
Alright. I have moved out the settings file version from the json to the file header so we can now check the version before trying to parse the json (in case it breaks again). Also, when settings file version changes a suggestion to re-export settings file will be shown to the user so they can keep their settings as updated as possible (probably nobody will do that, but just in case).

Adamantcheese commented on Clover/app/src/main/java/org/floens/chan/ui/controller/export/ImportExportSettingsController.java:
```diff
@@ -0,0 +1,322 @@
+                .setTitle(context.getString(R.string.clover_directory_may_not_exist_title))
+                .setMessage(context.getString(R.string.clover_directory_may_not_exist_message))
+                .setPositiveButton(context.getString(R.string.create), (dialog1, which) -> {
+                    getPermissionHelper().requestPermission(Manifest.permission.READ_EXTERNAL_STORAGE, granted -> {
```
This should be WRITE_EXTERNAL_STORAGE, because we're making the directories here which is writing, not reading.

K1rakishou replied:
Huh, that's funny, because couple of lines above I'm checking whether we already have WRITE_EXTERNAL_STORAGE and if not then for some reason requesting the READ_EXTERNAL_STORAGE.
